[REFACTOR] EXPIRED_JWT_TOKEN 분리

jnujh opened the pull request:
## 🚀 작업 내용

프론트엔드 개발자의 요청에 따라, 클라이언트에서 인증 상태를 더 명확하게 처리할 수 있도록 토큰 만료 관련 에러 응답을 세분화했습니다.

- **에러 코드 분리 및 명확화**
    - 기존의 범용적인 **EXPIRED_JWT_TOKEN** 에러 코드를 폐기했습니다.
    - Access Token 만료 시에는 **EXPIRED_ACCESS_TOKEN**을, Refresh Token 만료 시에는 **EXPIRED_REFRESH_TOKEN**을 반환하도록 **ErrorCode Enum**을 수정했습니다.
- **Refresh Token 만료 처리 로직 수정 (AuthServiceImpl)**
    - 토큰 재발급(reissue) 로직 내에서 Refresh Token이 만료되었을 때 발생하는 **ExpiredJwtException**을 **try-catch**로 잡아, **CustomException(ErrorCode.EXPIRED_REFRESH_TOKEN)**을 던지도록 수정했습니다.
- **Access Token 만료 처리 로직 수정 (JwtAuthenticationFilter)**
    - 모든 API 요청의 진입점인 **JwtAuthenticationFilter**에서 Access Token 만료 시 **EXPIRED_ACCESS_TOKEN** 에러 코드를 반환하도록 수정했습니다.
- **테스트 코드 추가 및 수정**
    - **AuthControllerTest**: Refresh Token이 만료되었을 때, 재발급 API가 의도한 대로 401 Unauthorized 상태와 **EXPIRED_REFRESH_TOKEN** 에러 코드를 반환하는지 검증하는 테스트 케이스를 추가했습니다.
    - **JwtAuthenticationFilterTest**: Access Token 만료 시 반환되는 에러 메시지가 새로운 에러 코드의 메시지와 일치하도록 테스트 검증 로직을 수정했습니다.

추가로 reissue 재발급 로직이 무한루프가 도는 문제를 해결하기 위해 
JwtAuthenticationFilter 이중 등록을 제거하고 
WebSecurityCustomizer ignoring()을 사용했습니다.

---

## ⏱️ 소요 시간

2h

---

## 🤔 고민했던 내용

1. **문제의 시작: 프론트엔드의 요청**
    - 처음에는 Access Token과 Refresh Token이 만료되었을 때 모두 동일한 **EXPIRED_JWT_TOKEN** 에러 코드를 반환하고 있었습니다.
    - 이로 인해 프론트엔드에서는 두 가지 상황을 구분할 수 없어, 다음과 같은 후속 처리에 어려움이 있었습니다.
        - **Access Token 만료:** Refresh Token을 이용해 Access Token 재발급을 요청해야 함.
        - **Refresh Token 만료:** 세션이 완전히 끝났으므로, 사용자를 로그인 페이지로 보내야 함.
    - 이 두 시나리오의 클라이언트 처리 방식이 완전히 다르기 때문에, 서버에서 내려주는 에러 코드를 분리하는 것이 필수적이었습니다.
2. **예외의 명확한 분리와 중앙 관리**
    1. **1단계: 에러 코드부터 명확하게 정의**
        - 가장 먼저 **ErrorCode Enum**에 **EXPIRED_ACCESS_TOKEN**과 **EXPIRED_REFRESH_TOKEN**을 명시적으로 정의했습니다. 
    2. **2단계: 각 토큰이 만료되는 '장소'에서 예외를 처리**
        - **Access Token:** 거의 모든 인증이 필요한 API 요청의 가장 앞단인 **JwtAuthenticationFilter**에서 검증됩니다. 따라서 이 필터에서 만료 예외가 발생했을 때 **EXPIRED_ACCESS_TOKEN**을 반환하도록 수정하는 것이 가장 적절하다고 판단했습니다.
        - **Refresh Token:** 오직 토큰 재발급(reissue) 로직에서만 사용됩니다. 따라서 **AuthServiceImpl**의 **reissue** 메소드 내부에서 만료 예외를 처리하는 것이 책임과 역할 분리에 가장 적합하다고 판단했습니다. **try-catch**를 사용해 jjwt 라이브러리가 던지는 **ExpiredJwtException**을 정의된 **CustomException**으로 '전환'하여, 비즈니스에 더 가까운 예외를 던지도록 구현했습니다.
    3. **3단계: 변경 사항을 보장하는 테스트를 작성**
        - 새로 추가된 Refresh Token 만료 시나리오에 대한 컨트롤러 테스트(**AuthControllerTest**)를 작성하여, 실제 API 호출 시 의도한 JSON 응답이 반환되는지 전체 흐름을 검증했습니다. 이를 통해 향후 리팩토링 시에도 해당 기능이 깨지지 않음을 보장할 수 있습니다.

---

## 💬 리뷰 중점사항

- **AuthServiceImpl**의 **reissue** 메소드에서 **try-catch**를 사용하여 예외를 전환하는 방식이 적절한지 확인 부탁드립니다. Refresh Token 만료라는 비즈니스 상황을 서비스 계층에서 명확히 표현하기 위한 선택이었습니다.
- **JwtAuthenticationFilter**에서 Access Token 만료를 처리하고, **AuthServiceImpl**에서 Refresh Token 만료를 처리하는 역할 분리가 명확하게 이루어졌는지 검토 부탁드립니다.
- 새로 추가된 **reissue_fail_expiredRefreshToken** 테스트 케이스가 Refresh Token 만료 시나리오를 충분히 잘 검증하고 있는지 확인 부탁드립니다.

---

## 🔗 관련 이슈

- Close #172

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **버그 수정**
  * 만료된 토큰 처리 분리: Access/Refresh 토큰 만료 상황에 대해 구체적인 오류 코드 및 메시지로 응답
  * 리프레시 토큰 만료 시 재로그인 안내 반환

* **보안/인증**
  * 재발급 엔드포인트 재요청 시 인증 필터 우회 처리 추가
  * 만료 토큰 예외 매핑과 JWT 필터 통합 로직 개선

* **테스트**
  * 테스트 및 API 입력에서 전화번호 형식을 하이픈 표준으로 통일
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jnujh committed:
feat: 토큰 만료 에러 코드 세분화 (#172)

Access Token 만료와 Refresh Token 만료를 구분하기 위해 에러 코드를 분리합니다.
•Access Token 만료: EXPIRED_ACCESS_TOKEN
•Refresh Token 만료: EXPIRED_REFRESH_TOKEN

이를 통해 클라이언트에서 각 에러에 따른 분기 처리가 가능해집니다.

jnujh committed:
fix: 전화번호 형식 수정에 따른 테스트코드 수정

jnujh committed:
Feat: Refresh Token 만료 시 전용 에러 코드 반환 (#172)

기존에는 Access Token과 Refresh Token 만료 시 동일한 예외가 발생하여 클라이언트에서 두 상황을 구분하기 어려웠습니다.

토큰 재발급(`reissue`) 로직에서 Refresh Token 검증 중 `ExpiredJwtException`이 발생할 경우,
이를 `CustomException(ErrorCode.EXPIRED_REFRESH_TOKEN)`으로 전환하여 던지도록 수정했습니다.

이를 통해 클라이언트는 Refresh Token이 만료되었음을 명확히 인지하고, 사용자에게 재로그인을 안내하는 등의 후속 조치를 할 수 있습니다.

jnujh committed:
Refactor: Access Token 만료 시 EXPIRED_ACCESS_TOKEN 에러 반환 (#172)

토큰 만료 에러 코드를 세분화하는 작업의 일환으로, `JwtAuthenticationFilter`를 수정했습니다.

요청의 Access Token이 만료되어 `ExpiredJwtException`이 발생했을 때,
기존의 범용적인 에러 코드 대신 명시적인 `ErrorCode.EXPIRED_ACCESS_TOKEN`을 사용하도록 변경했습니다.

이 변경으로 클라이언트는 Access Token 만료 상황을 명확하게 식별하고, 토큰 재발급 플로우를 트리거할 수 있습니다.

jnujh committed:
Test: Refresh Token 만료 시 재발급 실패 테스트 추가 (#172)

토큰 재발급 API(`/api/auth/reissue`) 호출 시, Refresh Token이 만료된 경우에 대한 컨트롤러 테스트 케이스를 추가합니다.

AuthService가 `CustomException(ErrorCode.EXPIRED_REFRESH_TOKEN)`을 던지는 상황을 모킹하고,
API 호출 결과가 예상대로 HTTP 401 상태 코드와 `EXPIRED_REFRESH_TOKEN` 에러 코드를 포함하는지 검증합니다.

jnujh committed:
Test: Access Token 만료 에러 코드 변경에 따른 테스트 수정 (#172)

`JwtAuthenticationFilter`에서 Access Token 만료 시 반환하는 에러 코드가 `EXPIRED_ACCESS_TOKEN`으로 변경됨에 따라, 관련 통합 테스트 코드를 수정합니다.

만료된 토큰으로 API 요청 시, 응답 JSON의 `message` 필드가
`ErrorCode.EXPIRED_ACCESS_TOKEN`에 정의된 메시지와 일치하는지 검증하도록 `jsonPath` 단언문을 업데이트했습니다.

jnujh committed:
Refactor: JwtAuthenticationFilter의 자동 등록 방지

기존에는 @Component 어노테이션으로 인해 JwtAuthenticationFilter가 Spring의 컴포넌트 스캔 과정에서 자동으로 Bean으로 등록되었습니다.
이로 인해 SecurityConfig의 ignoring() 설정과 무관하게 필터가 동작하여, 필터가 이중으로 등록되는 문제가 발생했습니다.

이 문제를 해결하고 필터의 생성 및 등록 시점을 명확하게 제어하기 위해 @Component 어노테이션을 제거합니다.

이제 JwtAuthenticationFilter는 SecurityConfig 내에서 수동으로 생성되고 필터 체인에 등록되어,
정확히 한 번만, 그리고 의도된 보안 규칙에 따라서만 동작하게 됩니다.

jnujh committed:
Fix: 필터 수동 등록 및 ignoring() 설정으로 reissue 무한 루프 해결

토큰 재발급(/api/auth/reissue) 요청 시, 만료된 Access Token으로 인해 인증 필터에서 차단되어 무한 루프가 발생하는 문제를 해결합니다.

1.  WebSecurityCustomizer 추가:
    -   /api/auth/** 등 인증이 필요 없는 경로를 Spring Security 필터 체인에서 완전히 제외(ignore)하도록 설정했습니다.
    -   이를 통해 /api/auth/reissue 요청이 헤더에 만료된 토큰을 포함하더라도, 인증 필터를 거치지 않고 컨트롤러까지 안전하게 도달할 수 있습니다.

2.  JwtAuthenticationFilter 수동 등록:
    - 기존의 필터 이중 등록 문제를 해결하기 위해, SecurityConfig가 JwtAuthenticationFilter를 직접 생성(new)하고 필터 체인에 등록하도록 변경했습니다.
    - 이를 위해 생성자를 통해 필요한 의존성(JwtProvider 등)을 주입받도록 구조를 개선했습니다.

jnujh committed:
Test: /api/auth/** 경로에 대한 인증 필터 무시 검증 테스트 추가

SecurityConfig에 추가된 WebSecurityCustomizer의 ignoring() 설정이 올바르게 동작하는지 검증하는 통합 테스트 케이스를 추가합니다.

테스트 시나리오는 다음과 같습니다.
- 일부러 만료된 Access Token을 헤더에 담아 /api/auth/reissue 경로로 요청합니다.
- 인증 필터가 무시되었다면 요청은 401(Unauthorized)이 아닌, 컨트롤러 레벨의 에러(400 Bad Request)를 반환해야 합니다.

ji-mim commented on src/main/java/com/kakaotech/team18/backend_server/global/config/SecurityConfig.java:
```diff
@@ -23,13 +26,29 @@
-    public SecurityFilterChain filterChain(HttpSecurity http, JwtAuthenticationFilter jwtAuthenticationFilter) throws Exception {
+    public WebSecurityCustomizer webSecurityCustomizer() {
+        return (web) -> web.ignoring()
+                .requestMatchers("/api/auth/**", "/swagger-ui.html", "/v3/api-docs/**", "/swagger-ui/**");
```
혹시 h2 콘솔도 추가해주실 수 있으실까요??

jnujh replied:
추가했습니다 !

ji-mim commented on src/main/java/com/kakaotech/team18/backend_server/domain/auth/service/AuthServiceImpl.java:
```diff
@@ -191,7 +194,14 @@ public LoginSuccessResponseDto register(String bearerToken, RegisterRequestDto r
+        } catch (ExpiredJwtException e) {
+            // Refresh Token이 만료된 경우, 새로운 에러 코드로 예외 발생
+            throw new CustomException(ErrorCode.EXPIRED_REFRESH_TOKEN);
+        }
```
혹시 에러를 추가로 만들지 않고 CustomException을 던지시는 건 굳이 그렇게까지 할 필요가 없다고 생각하셨기 때문일까요?

jnujh replied:
앗 제가 이 부분 놓쳤네요 !
수정하겠습니다

jnujh replied:
수정했습니다

ji-mim reviewed (approved):
1. jwt의 토큰이 만료되는 경우를 엑세스 토큰과 리프레쉬 토큰이 만료되는 경우로 분할해서 프론트가 이에 맞는 대응을 하도록 함 
2. JWT Auth 필터를 자동 등록이 아니라 필터가 동작할 떄 수동으로 등록하도록 해 ignore된 API에는 정상적인 요청이 갈 수 있도록 함 

이정도가 주된 변경 사항이라고 볼 수 있을까요? 
자동 빈 등록 때문에 계속 JWTAuth 필터가 동작하고 있다는 생각을 전혀 하지 못했는데 이렇게 배워가네요..
저번 카카오 로그인 때 만료된 토큰이라고 뜬건 프론트 입장에서 어떤 토큰이 만료됐는지 알 수 없었고, 토큰 재발급 요청이 jwt 필터 때문에 이루어지지 않았던 것 때문일까요? 
엑세스 토큰이 로컬 세션..?에 저장됐던건 어떻게 됐나요? 

수고 많으셨습니다!!

jnujh commented:
엑세스 토큰이 로컬 세션..?에 저장됐던건 어떻게 됐나요?

정확히 어떤 부분을 말씀하시는지 모르겠어요 ㅠㅠ

jnujh committed:
refactor: h2콘솔 web.ignoring 추가

jnujh committed:
refactor: RefreshToken 만료에 대핸 구체적인 예외 클래스 생성

ji-mim commented:
@jnujh 
> 엑세스 토큰이 로컬 세션..?에 저장됐던건 어떻게 됐나요?
> 
> 정확히 어떤 부분을 말씀하시는지 모르겠어요 ㅠㅠ
10월 28일 회의록을 참고해주시면 좋을 것 같아요!
<img width="1390" height="506" alt="CleanShot 2025-11-01 at 21 08 34@2x" src="https://github.com/user-attachments/assets/60f33d79-a442-4db3-8890-c9d3561ca231" />

저번에 토큰이 잘못된 저장소..?에 저장된걸로 기억해서 여쭤봤어요!

jnujh commented:
액세스토큰은 로컬 스토리지에 
임시토큰은 세션 스토리지에 저장되는게 맞습니다 !

jnujh commented on src/main/java/com/kakaotech/team18/backend_server/global/config/SecurityConfig.java:
```diff
@@ -23,17 +27,34 @@
 
+    /**
+     * 정적 리소스나 인증이 전혀 필요 없는 경로들을 Spring Security 필터 체인에서 제외합니다.
+     */
     @Bean
-    public SecurityFilterChain filterChain(HttpSecurity http, JwtAuthenticationFilter jwtAuthenticationFilter) throws Exception {
+    public WebSecurityCustomizer webSecurityCustomizer() {
+        return (web) -> web.ignoring()
+                .requestMatchers("/h2-console/**", "/api/auth/**", "/swagger-ui.html", "/v3/api-docs/**", "/swagger-ui/**");
+    }
```
문제의 시작: 끝나지 않는 재발급 요청
가장 처음 마주한 문제는 "Access Token 만료 시, 토큰 재발급 요청이 무한 반복되는 현상" 이었습니다.
프론트엔드에서는 Access Token이 만료되었다는 401 에러를 받으면, 재발급을 위해 /api/auth/reissueAPI를 호출합니다. 하지만 이 재발급 요청조차도 동일한 401 에러를 받으면서 무한 루프에 빠졌습니다.
첫 번째 가설: permitAll()이 필터를 막지 못한다
처음에는 SecurityConfig에 설정된 permitAll() 규칙이 JwtAuthenticationFilter의 동작을 막지 못한다고 생각했습니다.
이 가설은 맞았습니다. **permitAll()**은 인가(Authorization) 단계에서 권한을 허용하는 것이고, JwtAuthenticationFilter는 그보다 앞선 인증(Authentication) 단계에서 동작하기 때문입니다.
그래서 해결책으로 WebSecurityCustomizer의 ignoring() 설정을 도입했습니다. **ignoring()**은 필터 체인 자체를 우회하므로, /api/auth/reissue 요청이 인증 필터를 거치지 않고 컨트롤러까지 도달할 것이라고 예상했습니다.
두 번째 난관: 테스트의 배신
ignoring() 설정을 추가하고 검증 테스트를 작성했지만, 테스트는 여전히 실패했습니다.
기대했던 400 Bad Request(컨트롤러 에러) 대신, 계속해서 401 Unauthorized(인증 필터 에러)가 발생했습니다.
이 결과는 "우리가 추가한 ignoring() 설정이 적용되지 않았거나, 다른 무언가가 여전히 필터를 동작시키고 있다" 는 강력한 증거였습니다.
진짜 원인의 발견: 이중으로 등록된 필터
여기서부터 Spring Security의 내부 동작과 Bean 관리 방식에 대해 깊이 파고들었습니다. 문제의 진짜 원인은 JwtAuthenticationFilter가 두 번 등록되고 있었다는 점이었습니다.
자동 등록: JwtAuthenticationFilter 클래스에 붙어있던 https://github.com/component 어노테이션 때문에, Spring이 이 필터를 자동으로 Bean으로 만들고 필터 체인에 등록했습니다. 이 '자동 등록된 필터'는 SecurityConfig의 ignoring() 설정을 알지 못했습니다.
수동 등록: SecurityConfig의 filterChain 메소드에서 **http.addFilterBefore()**를 통해 우리가 의도한 필터를 등록하고 있었습니다.
결국, /api/auth/reissue 요청이 들어왔을 때, 우리가 의도한 '수동 등록 필터'는 ignoring() 설정에 따라 동작하지 않았지만, 그보다 먼저 '자동 등록된 필터'가 요청을 가로채서 만료된 토큰을 보고 401 에러를 발생시켰던 것입니다.
최종 해결책: 필터 제어권의 완전한 장악
원인을 파악한 후 해결책은 명확해졌습니다. 필터의 등록을 우리가 완벽하게 제어해야 했습니다.
https://github.com/component 제거: JwtAuthenticationFilter에서 https://github.com/component를 제거하여 자동 등록의 여지를 없앴습니다.
수동 생성 및 등록: SecurityConfig가 JwtAuthenticationFilter를 만드는 데 필요한 모든 의존성(JwtProvider등)을 직접 주입받은 뒤, new 키워드로 필터를 생성하여 filterChain에 등록하도록 변경했습니다.
이 변경을 통해 JwtAuthenticationFilter는 정확히 한 번만, 그리고 우리가 정의한 보안 규칙(ignoring() 포함)을 완벽하게 따르는 방식으로만 동작하게 되었습니다. 이로써 무한 루프 문제가 해결되었고, 검증 테스트도 성공적으로 통과할 수 있었습니다.
이러한 이유로 의도하고 작성한건데 어떻게 생각해

jnujh replied:
수정했거든? 확인해볼래?

jnujh committed:
fix: JWT 필터가 토큰 재발급 요청을 차단하는 문제 해결

## 문제 원인

만료된 액세스 토큰을 헤더에 담아 토큰 재발급 API(`/api/auth/reissue`)를 호출할 때,
`JwtAuthenticationFilter`가 컨트롤러에 도달하기 전에 만료된 토큰을 먼저 검증하여
`ExpiredJwtException`을 발생시키는 문제가 있었습니다.

이로 인해 클라이언트는 정상적으로 토큰을 재발급받지 못하고 인증 오류(401 Unauthorized)를 응답받게 되었습니다.

기존에는 이 문제를 해결하기 위해 `SecurityConfig`에서 `ignoring()` 설정을 사용하여
`/api/auth/**` 경로를 Spring Security 필터 체인에서 완전히 제외했습니다.
하지만 이 방식은 해당 경로에 CORS, 보안 헤더 등 다른 보안 설정들이 적용되지 않는 부작용이 있었습니다.

## 해결 방안

이 문제를 근본적으로 해결하고 보안 일관성을 유지하기 위해 다음과 같이 수정했습니다.

1.  **`SecurityConfig.java` 수정**
    *   `ignoring()` 목록에서 `/api/auth/**` 경로를 제거했습니다.
    *   이제 모든 API 요청이 Spring Security 필터 체인을 통과하게 되어, 모든 경로에 일관된 보안 정책(CORS, 보안 헤더 등)이 적용됩니다.

2.  **`JwtAuthenticationFilter.java` 수정**
    *   `doFilterInternal` 메서드 시작 부분에 요청 URI를 확인하는 로직을 추가했습니다.
    *   요청 URI가 `/api/auth/`로 시작하는 경우, 토큰 검증 로직을 건너뛰고 즉시 다음 필터로 요청을 전달하도록 수정했습니다.
    *   이를 통해 토큰 재발급 요청이 만료된 토큰의 영향을 받지 않고 안전하게 컨트롤러에 도달할 수 있게 되었습니다.
